app/src/main/java/com/spascoding/taskycourse/feature_auth/data/repository/AuthRepositoryImpl.kt

```diff
@@ -54,4 +60,63 @@ class AuthRepositoryImpl @Inject constructor(
         return response.isSuccessful
     }
 
+    //Utils
+
+    private suspend fun <T> performRequest(
+        request: suspend () -> Response<T>,
+        onSuccess: suspend (T) -> Unit = {}
+    ): Result<T, DataError.Remote> {
+        return try {
+            val response = request()
+            if (response.isSuccessful) {
+                response.body()?.let {
+                    onSuccess(it)
+                    Result.Success(it)
+                } ?: Result.Error(DataError.Remote.UNKNOWN)
+            } else {
+                //How I can return error massage from server and it needs to do this?
+//                val errorResponse = response.errorBody()?.string()
+//                val errorMessage = handleErrorMassage(errorResponse)
+//                Result.Error(TextError(errorMessage))
+                Result.Error(DataError.Remote.UNKNOWN)
+            }
+        } catch (e: HttpException) {
+            handleErrorCode(e.code())
+        } catch (e: IOException) {
+            Result.Error(DataError.Remote.NO_INTERNET)
+        }
+    }
+
+    private fun handleErrorMassage(errorResponse: String?): String {
+        return if (errorResponse != null) {
+            try {
+                val jsonObject = JSONObject(errorResponse)
+                jsonObject.getString("message")
+            } catch (e: Exception) {
+                "Unknown error"
+            }
+        } else {
+            "Unknown error"
+        }
+    }
+
+    private fun handleErrorCode(code: Int): Result.Error<DataError.Remote> {
+        return when (code) {
+            400 -> Result.Error(DataError.Remote.BAR_REQUEST)
+            401 -> Result.Error(DataError.Remote.UNAUTHORIZED)
+            403 -> Result.Error(DataError.Remote.FORBIDDEN)
+            404 -> Result.Error(DataError.Remote.NOT_FOUND)
+            405 -> Result.Error(DataError.Remote.METHOD_NOT_ALLOWED)
+            408 -> Result.Error(DataError.Remote.REQUEST_TIMEOUT)
+            413 -> Result.Error(DataError.Remote.PAYLOAD_TOO_LARGE)
+
+            500 -> Result.Error(DataError.Remote.INTERNAL_SERVER_ERROR)
+            502 -> Result.Error(DataError.Remote.BAD_GATEWAY)
+            503 -> Result.Error(DataError.Remote.SERVICE_UNAVAILABLE)
+            504 -> Result.Error(DataError.Remote.GATEWAY_TIMEOUT)
```

Comment:
Is this distinction really something the average user would benefit from? All these types of server errors rather hint towards showing a generic `Oops, something went wrong` error.